src/toast/ops/filterbin.py

```diff
@@ -0,0 +1,1098 @@
+        )
+
+        self._load_deprojection_map(data)
+        log.info_rank(
+            "OpFilterBin: Loaded deprojection map in", comm=self.comm, timer=timer
+        )
+
+        self._bin_map(data, detectors, filtered=False)
+        log.info_rank(
+            "OpFilterBin: Binned unfiltered map in", comm=self.comm, timer=timer
+        )
+
+        log.info_rank("OpFilterBin: Filtering signal", comm=self.comm)
+
+        timer1 = Timer()
+        timer1.start()
+        timer2 = Timer()
+        timer2.start()
+
+        t1 = time()
+        for iobs, obs in enumerate(data.obs):
+            dets = obs.select_local_detectors(detectors)
+            if self.grank == 0:
+                log.verbose(
+                    f"{self.group:4} : OpFilterBin: Processing observation "
+                    f"{iobs} / {len(data.obs)}",
+                )
+
+            common_flags = obs.shared[self.shared_flags].data
+            phase = self._get_phase(obs)
+            common_templates = self._build_common_templates(obs, phase, common_flags)
+
+            if self.grank == 0:
+                log.verbose(
+                    f"{self.group:4} : OpFilterBin:   Built common templates in "
+                    f"{time() - t1:.2f} s",
+                )
+                t1 = time()
+
+            for idet, det in enumerate(dets):
+                if self.grank == 0:
+                    log.verbose(
+                        f"{self.group:4} : OpFilterBin:   Processing detector "
+                        f"# {idet + 1} / {len(dets)}",
+                    )
+
+                signal = obs.detdata[self.det_data][det]
+                flags = obs.detdata[self.det_flags][det]
+                # `good` is essentially the diagonal noise matrix used in
+                # template regression.  All good detector samples have the
+                # same noise weight and rest have zero weight.
+                good_fit = np.logical_and(
+                    (common_flags & self.shared_flag_mask) == 0,
+                    (flags & self.det_flag_mask) == 0,
+                )
+                good_bin = np.logical_and(
+                    (common_flags & self.binning.shared_flag_mask) == 0,
+                    (flags & self.binning.det_flag_mask) == 0,
+                )
+                if np.sum(good_fit) == 0:
+                    continue
+
+                deproject = (
+                    self.deproject_map is not None
+                    and self._deproject_pattern.match(det) is not None
+                )
+
+                if deproject or self.write_obs_matrix:
+                    # We'll need pixel numbers
+                    obs_data = data.select(obs_uid=obs.uid)
+                    self.binning.pixel_pointing.apply(obs_data, detectors=[det])
+                    pixels = obs.detdata[self.binning.pixel_pointing.pixels][det]
+                    # and weights
+                    self.binning.stokes_weights.apply(obs_data, detectors=[det])
```

Comment:
So in this scenario, we just assume that we job has enough memory to expand pixels and weights and leave them in memory in each observation?  Or do we check whether they exist and if not, create them here and delete them after each observation?

Reply:
I see now that you are calling `apply()` on only a single detector and a single observation.  This is fine in the case of operators like the pixels and weights operators that do not have any operations in their `_finalize()` method, and which currently only have a CPU implementation.  In the general case, one would use a `Pipeline` that used a list of operators composed of pixels, weights, and the bulk of the code below (implemented as a another operator).  The Pipeline calls exec for each operator in turn and runs finalize at the end.  This change can wait until we have GPU kernels for those 3 things.